src/typechecker.erl

```diff
@@ -653,18 +644,18 @@ negate_num_type({type, _, TyName, []} = Ty) when
     Ty;
 negate_num_type({integer, P, I}) ->
     {integer, P, -I};
-negate_num_type(IntTy) ->
-    {L, U} = int_type_to_range(IntTy),
-    L2 = case U of
-             pos_inf -> neg_inf;
-             _ -> -U
-         end,
-    U2 = case L of
-             neg_inf -> pos_inf;
-             _ -> -L
-         end,
-    normalize({type, erl_anno:new(0), union, int_range_to_types({L2, U2})},
-              #tenv{}).
+negate_num_type(RangeTy) ->
+    %% some kind of range type like `1..3' or `neg_integer()'
+    {L, U} = int_type_to_range(RangeTy),
+    L2 = int_negate(U),
+    U2 = int_negate(L),
+    case int_range_to_types({L2, U2}) of
+        [Ty] ->
+            Ty;
+        Tys = [_, _] ->
+            %% in some cases the result is two mutually exclusive type
+            {type, erl_anno:new(0), union, Tys}
```

Comment:
The only case I can think of where a non-union results in a union is the negation of `non_neg_integer()` => `neg_integer() | 0`. Are there any other cases?

How about negating a union of int types such as `neg_integer() | 0` or `1..3 | 5..7`? That would be possible as well, just negate each term in the union and normalize again. That's a future improvement perhaps.

Reply:
you are right about non-union=>union.

and good point about union of integer types. actually they currently crash, so I'll try to address that case too.